cpp/src/arrow/builder.h

```diff
@@ -91,30 +91,31 @@ class ARROW_EXPORT ArrayBuilder {
   /// Set the next length bits to not null (i.e. valid).
   Status SetNotNull(int64_t length);
 
-  /// Allocates initial capacity requirements for the builder.  In most
-  /// cases subclasses should override and call their parent class's
-  /// method as well.
-  virtual Status Init(int64_t capacity);
+  /// \brief Ensure that at least the indicated number of elements can be appended to
+  /// the builder without a reallocation. Does not account for reallocations
+  /// that may be due to variable size data, like binary values.
+  /// \param[in] capacity the minimum number of additional array values
+  /// \return Status
+  virtual Status Resize(int64_t capacity);
 
-  /// Resizes the null_bitmap array.  In most
-  /// cases subclasses should override and call their parent class's
-  /// method as well.
-  virtual Status Resize(int64_t new_bits);
+  /// \brief Ensure that there is enough space allocated to add the indicated
+  /// number of elements without any further calls to Resize
+  /// \param[in] additional_capacity the number of additional array values
+  /// \return Status
+  Status Reserve(int64_t additional_capacity);
 
   /// Reset the builder.
   virtual void Reset();
 
-  /// Ensures there is enough space for adding the number of elements by checking
-  /// capacity and calling Resize if necessary.
-  Status Reserve(int64_t elements);
-
   /// For cases where raw data was memcpy'd into the internal buffers, allows us
   /// to advance the length of the builder. It is your responsibility to use
   /// this function responsibly.
   Status Advance(int64_t elements);
 
   ARROW_DEPRECATED("Use Finish instead")
-  std::shared_ptr<PoolBuffer> null_bitmap() const { return null_bitmap_; }
+  std::shared_ptr<PoolBuffer> null_bitmap() const {
```

Comment:
Perhaps bite the bullet and return a `ResizableBuffer` here? `PoolBuffer` doesn't have any additional methods anyway.

Reply:
done